### PR TITLE
cmgd/lib: cmgd performance improvements

### PR DESCRIPTION
**Pull request by rameshabhinay**
This commit includes changes to improve performance
of set config and also adds show commands to record
set-cfg performance and debug logs to record apply-cfg
performance.

Signed-off-by: Abhinay Ramesh <rabhinay@vmware.com>

### CODE REVIEW

File: cmgd/cmgd_bcknd_adapter.h [hunk 1]
```diff
@@ -94,6 +94,13 @@ typedef void (*cmgd_bcknd_get_data_result_notify_t)(
 typedef void (*cmgd_bcknd_data_notify_t)(
         struct nb_yang_xpath *xpaths[], int num_xpaths);
 
+typedef struct cmgd_bcnkd_apply_stats_ {
```
Comment (pushpasis) [inline cmgd/cmgd_bcknd_adapter.h:97]: Not necessary. Why is this on the backend adapter?

File: cmgd/cmgd_frntnd_adapter.c [hunk 2]
```diff
@@ -195,6 +195,21 @@ static void cmgd_frntnd_sessn_show_trxn_cleanup(cmgd_frntnd_sessn_ctxt_t *sessn)
...
+		set_cfg_stats->max_tm = set_cfg_stats->last_exec_tm;
+
+	if (set_cfg_stats->last_exec_tm < set_cfg_stats->min_tm)
+		set_cfg_stats->min_tm = set_cfg_stats->last_exec_tm;
+
+	set_cfg_stats->avg_tm = (set_cfg_stats->avg_tm +
+					set_cfg_stats->last_exec_tm)/2;
+}
+
 static void cmgd_compute_timers(cmgd_sessn_commit_stats_t *cmt_stats)
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:213]: Rename this to cmd_compute_commit_timers()

File: cmgd/cmgd_frntnd_adapter.c [hunk 3]
```diff
@@ -413,7 +429,9 @@ static int cmgd_frntnd_send_setcfg_reply(cmgd_frntnd_sessn_ctxt_t *sessn,
 
 	CMGD_FRNTND_ADPTR_DBG("Sending SET_CONFIG_REPLY message to CMGD Frontend client '%s'",
 		sessn->adptr->name);
-
+	cmgd_get_realtime(&sessn->adptr->set_cfg_stats.end,
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:432]: Rewrite this as... 
if (set_cfg_stats.enable)
    cmd_ger_realtime(&sessn->adptr->set_cfg_stats.end);

Comment (pushpasis) [reply cmgd/cmgd_frntnd_adapter.c:432]: Also why is 'set_cfg_stats.enable' recorded per session? This should be stored under cmd_master{}.

File: lib/cmgd_bcknd_client.h [hunk 4]
```diff
@@ -37,11 +37,13 @@ extern "C" {
...
-#define CMGD_BCKND_MSG_PROC_DELAY_MSEC		5
-#define CMGD_BCKND_MAX_NUM_MSG_PROC		100
+#define CMGD_BCKND_MSG_PROC_DELAY_MSEC		1
+#define CMGD_BCKND_MSG_PROC_DELAY_USEC		5
+#define CMGD_BCKND_MAX_NUM_MSG_PROC		500
 
-#define CMGD_BCKND_MSG_WRITE_DELAY_MSEC		50
-#define CMGD_BCKND_MAX_NUM_MSG_WRITE		1000
+#define CMGD_BCKND_MSG_WRITE_DELAY_MSEC		1
+#define CMGD_BCKND_MSG_WRITE_DELAY_USEC		5
```
Comment (pushpasis) [inline lib/cmgd_bcknd_client.h:45]: Too less..

File: lib/cmgd_frntnd_client.h [hunk 5]
```diff
@@ -43,11 +43,13 @@ extern "C" {
...
-#define CMGD_FRNTND_MSG_PROC_DELAY_MSEC			5
-#define CMGD_FRNTND_MAX_NUM_MSG_PROC			100
+#define CMGD_FRNTND_MSG_PROC_DELAY_MSEC			1
+#define CMGD_FRNTND_MSG_PROC_DELAY_USEC			5
+#define CMGD_FRNTND_MAX_NUM_MSG_PROC			500
 
-#define CMGD_FRNTND_MSG_WRITE_DELAY_MSEC		50
-#define CMGD_FRNTND_MAX_NUM_MSG_WRITE			1000
+#define CMGD_FRNTND_MSG_WRITE_DELAY_MSEC		1
+#define CMGD_FRNTND_MSG_WRITE_DELAY_USEC		5
```
Comment (pushpasis) [inline lib/cmgd_frntnd_client.h:51]: Again too less...

File: lib/cmgd_bcknd_client.h [hunk 6]
```diff
@@ -37,11 +37,13 @@ extern "C" {
...
+#define CMGD_BCKND_MAX_NUM_MSG_WRITE		500
```
Comment (pushpasis) [inline lib/cmgd_bcknd_client.h:46]: You could make this 100 instead.

File: cmgd/cmgd_frntnd_adapter.c [hunk 7]
```diff
@@ -255,6 +273,7 @@ static cmgd_frntnd_sessn_ctxt_t *cmgd_frntnd_create_session(
 	sessn->cfg_trxn_id = CMGD_TRXN_ID_NONE;
 	cmgd_frntnd_adapter_lock(adptr);
 	sessn->cmt_stats.min_tm = ULONG_MAX;
+	sessn->cmt_stats.enable = true;
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:276]: This should be global. Not per session.

Comment (pushpasis) [reply cmgd/cmgd_frntnd_adapter.c:276]: Please change this.

File: cmgd/cmgd_frntnd_adapter.c [hunk 8]
```diff
@@ -413,7 +434,9 @@ static int cmgd_frntnd_send_setcfg_reply(cmgd_frntnd_sessn_ctxt_t *sessn,
 
 	CMGD_FRNTND_ADPTR_DBG("Sending SET_CONFIG_REPLY message to CMGD Frontend client '%s'",
 		sessn->adptr->name);
-
+	if (sessn->adptr->set_cfg_stats.enable)
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:437]: Again should be global. Don't think should be per adapter.

File: cmgd/cmgd_frntnd_adapter.c [hunk 9]
```diff
@@ -1777,3 +1854,34 @@ void cmgd_frntnd_adapter_status_write(struct vty *vty, bool detail)
 	vty_out(vty, "  Total: %d\n", 
 		(int) cmgd_frntnd_adptr_list_count(&cmgd_frntnd_adptrs));
 }
+
+void cmgd_frntnd_adapter_perf_measurement(struct vty *vty, bool config)
+{
+	cmgd_frntnd_client_adapter_t *adptr;
+	cmgd_frntnd_sessn_ctxt_t *sessn;
+
+	FOREACH_ADPTR_IN_LIST(adptr) {
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:1863]: Same comment as before. Just have a global. Not per adapter. Maybe inside cmd_master.

File: cmgd/cmgd_frntnd_adapter.c [hunk 10]
```diff
@@ -1777,3 +1854,34 @@ void cmgd_frntnd_adapter_status_write(struct vty *vty, bool detail)
...
+	}
+}
+
+void cmgd_frntnd_adapter_reset_perf_stats(struct vty *vty)
+{
+	cmgd_frntnd_client_adapter_t *adptr;
+	cmgd_frntnd_sessn_ctxt_t *sessn;
+	bool enable;
+
+	FOREACH_ADPTR_IN_LIST(adptr) {
```
Comment (pushpasis) [inline cmgd/cmgd_frntnd_adapter.c:1877]: Same comment as above.

File: cmgd/cmgd_trxn.c [hunk 11]
```diff
@@ -2013,32 +2021,34 @@ static int cmgd_trxn_cleanup(struct thread *thread)
 static void cmgd_trxn_register_event(
 	cmgd_trxn_ctxt_t *trxn, cmgd_trxn_event_t event)
 {
+	struct timeval tv = {.tv_sec = 0, .tv_usec = CMGD_TRXN_PROC_DELAY_USEC};
```
Comment (pushpasis) [inline cmgd/cmgd_trxn.c:2024]: Please break this line.

File: cmgd/cmgd_trxn.h [hunk 12]
```diff
@@ -30,6 +30,7 @@
 #include "cmgd/cmgd_db.h"
 
 #define CMGD_TRXN_PROC_DELAY_MSEC               5
+#define CMGD_TRXN_PROC_DELAY_USEC        	10
```
Comment (pushpasis) [inline cmgd/cmgd_trxn.h:33]: Indentation issue.

File: cmgd/cmgd_vty.c [hunk 13]
```diff
@@ -198,6 +198,40 @@ DEFPY(show_cmgd_frntnd_adapter_detail,
...
+	CMGD_STR
+	"Enable performance measurement\n")
+{
+	cmgd_frntnd_adapter_perf_measurement(vty, true);
+
+	return CMD_SUCCESS;
+}
+
+DEFPY_HIDDEN(no_cmgd_performance_measurement,
+	no_cmgd_performance_measurement_cmd,
```
Comment (pushpasis) [inline cmgd/cmgd_vty.c:213]: I believe you can write a single function for both commands. Check with Yash.

File: lib/cmgd_bcknd_client.h [hunk 14]
```diff
@@ -37,11 +37,13 @@ extern "C" {
...
-#define CMGD_BCKND_MSG_PROC_DELAY_MSEC		5
-#define CMGD_BCKND_MAX_NUM_MSG_PROC		100
+#define CMGD_BCKND_MSG_PROC_DELAY_MSEC		1
+#define CMGD_BCKND_MSG_PROC_DELAY_USEC		10
+#define CMGD_BCKND_MAX_NUM_MSG_PROC		500
 
-#define CMGD_BCKND_MSG_WRITE_DELAY_MSEC		50
-#define CMGD_BCKND_MAX_NUM_MSG_WRITE		1000
+#define CMGD_BCKND_MSG_WRITE_DELAY_MSEC		1
+#define CMGD_BCKND_MSG_WRITE_DELAY_USEC		10
```
Comment (pushpasis) [inline lib/cmgd_bcknd_client.h:45]: Please increase this to at least 1msec

File: lib/cmgd_frntnd_client.h [hunk 15]
```diff
@@ -43,11 +43,13 @@ extern "C" {
...
-#define CMGD_FRNTND_MSG_PROC_DELAY_MSEC			5
-#define CMGD_FRNTND_MAX_NUM_MSG_PROC			100
+#define CMGD_FRNTND_MSG_PROC_DELAY_MSEC			1
+#define CMGD_FRNTND_MSG_PROC_DELAY_USEC			10
+#define CMGD_FRNTND_MAX_NUM_MSG_PROC			500
 
-#define CMGD_FRNTND_MSG_WRITE_DELAY_MSEC		50
-#define CMGD_FRNTND_MAX_NUM_MSG_WRITE			1000
+#define CMGD_FRNTND_MSG_WRITE_DELAY_MSEC		1
+#define CMGD_FRNTND_MSG_WRITE_DELAY_USEC		10
```
Comment (pushpasis) [inline lib/cmgd_frntnd_client.h:51]: Again make this 1ms at least

File: lib/cmgd_frntnd_client.h [hunk 16]
```diff
@@ -43,11 +43,13 @@ extern "C" {
...
+#define CMGD_FRNTND_MAX_NUM_MSG_WRITE			500
```
Comment (pushpasis) [inline lib/cmgd_frntnd_client.h:52]: And check with 100 for this.

**Review by pushpasis — COMMENTED** [review]
Please take care of these comments

File: lib/cmgd_bcknd_client.h (hunk 14 above)
Comment (rameshabhinay) [reply lib/cmgd_bcknd_client.h:45]: have replaced all write with CMGD_BCKND_MSG_WRITE_DELAY_MSEC which is 1msec

**Pull request merged**